### PR TITLE
Feature add ExpectedHours model

### PR DESCRIPTION
**Pull request by Merrique**

### CODE REVIEW

File: managers/models.py [hunk 1]
```diff
@@ -90,3 +96,52 @@ def change_suspended_if_project_has_stop_date(sender: Project, **kwargs: Any) ->
             When(Q(stop_date__lt=timezone.datetime.now()), then=Value(False)), default=kwargs["instance"].suspended
         )
     )
+
+
+class ExpectedHoursQuerySet(models.QuerySet):
+    def get_expected_hours_sum_for_a_date(self, for_date: date, excluded_id: Optional[int] = None) -> timedelta:
```
Comment (marsolekk) [inline managers/models.py:102]: In my opinion, the name of the parameter `for_date` can be simply changed to `date`.

**Review by marsolekk — COMMENTED** [review]
At the moment I've found only these mentioned issues. 
I wonder, branch name should contain a model name with uppercase letter and also is this `chore` or more like `feature`. Now these changes are invisible for users, but as you create new models, factories, you are working on some future feature, the name of the branch should indicate that. I am not sure about that, so please wait also for Karol(@kbeker) opinion about naming convention.

File: managers/factories.py [hunk 2]
```diff
@@ -10,3 +13,13 @@ class Meta:
...
+
+
+class ExpectedHoursFactory(factory.DjangoModelFactory):
+    class Meta:
+        model = ExpectedHours
+
+    date = factory.Faker("date")
+    user = factory.SubFactory("users.factories.UserFactory")
+    value = factory.LazyFunction(lambda: datetime.timedelta(hours=8))
+    project = factory.SubFactory("managers.factories.ProjectFactory")
```
[the remark this answers is not in the document]
Comment (Merrique) [reply managers/factories.py:25]: Added new commit for 'ExpectedHoursFactory'

File: managers/models.py [hunk 3]
```diff
@@ -90,3 +96,52 @@ def change_suspended_if_project_has_stop_date(sender: Project, **kwargs: Any) ->
...
+        if (
+            self.user.user_expected_hours.get_expected_hours_sum_for_a_date(self.date, self.pk) + self.value
+            > ExpectedHoursConstants.MAX_EXPECTED_HOURS.value
+        ):
+            raise ValidationError(message=ExpectedHoursStrings.EXPECTED_HOURS_EXCEED_24H_PER_DAY_PER_USER.value)
+        if not (
+            self.user.manager_projects.filter(pk=self.project.pk).exists()
+            or self.user.projects.filter(pk=self.project.pk).exists()
+        ):
+            raise ValidationError(message=ExpectedHoursStrings.USER_NOT_IN_THE_PROJECT.value)
```
[the remark this answers is not in the document]
Comment (Merrique) [reply managers/models.py:147]: Changed

File: managers/models.py [hunk 4]
```diff
@@ -90,3 +96,52 @@ def change_suspended_if_project_has_stop_date(sender: Project, **kwargs: Any) ->
...
+
+        if excluded_id is not None:
+            query_for_date = query_for_date.exclude(pk=excluded_id)
+
+        return query_for_date.aggregate(expected_hours_sum=Coalesce(models.Sum("value"), timedelta()))[
+            "expected_hours_sum"
+        ]
+
+
+class ExpectedHours(models.Model):
```
[the remark this answers is not in the document]
Comment (Merrique) [reply managers/models.py:113]: Changed

File: managers/tests/test_unit_model.py [hunk 5]
```diff
@@ -7,14 +7,146 @@
...
+    @parameterized.expand([(16,), (20,), (31,), (59,)])
+    def test_expected_hours_model_should_not_accept_value_which_is_not_divisible_by_15(self, minutes):
+        self.field_should_not_accept_input(
+            "value",
+            timedelta(minutes=minutes),
+            error_message=str(ExpectedHoursStrings.EXPECTED_HOURS_MINUTES_ARE_INCORRECT.value),
+        )
+
+    def test_expected_hours_model_sum_for_given_day_for_given_user_should_not_exceed_24(self):
+
```
[the remark this answers is not in the document]
Comment (Merrique) [reply managers/tests/test_unit_model.py:116]: Removed

File: managers/models.py (hunk 1 above)
Comment (Merrique) [reply managers/models.py:102]: Can't changed it. The change produces pylint error: Redefining name 'date' from outer scope.

File: managers/tests/test_unit_model.py [hunk 6]
```diff
@@ -7,14 +7,145 @@
...
+            name="Test Project", start_date=datetime.now().date(), stop_date=datetime.now().date() + timedelta(days=30)
+        )
+        self.project.save()
+        self.project.managers.add(self.user)
+        self.project.save()
+
+        self.required_input["user"] = self.user
+        self.required_input["project"] = self.project
+
+        self.SAMPLE_STRING_FOR_TYPE_VALIDATION_TESTS = "String for test"
```
Comment (kbeker) [inline managers/tests/test_unit_model.py:43]: You can write this variable using small case letters and just name it:
```python
self.incorrect_value = "incorrect value"
```

**Review by kbeker — CHANGES_REQUESTED** [review]
There are few things to be polished but in overall code looks nice. Good job!

**Comment by kbeker** [comment]
And of course this branch is feature not a chore. Please also fix it

File: managers/factories.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Merrique) [reply managers/factories.py:25]: Ok.

File: managers/models.py (hunk 1 above)
Comment (Merrique) [reply managers/models.py:69]: Changed

File: managers/models.py [hunk 7]
```diff
@@ -59,6 +65,55 @@ def clean(self) -> None:
             raise ValidationError(message=ProjectConstants.STOP_DATE_VALIDATION_ERROR_MESSAGE.value)
 
 
+class ExpectedHoursQuerySet(models.QuerySet):
```
[the remark this answers is not in the document]
Comment (Merrique) [reply managers/models.py:68]: changed

File: managers/models.py [hunk 8]
```diff
@@ -59,6 +65,55 @@ def clean(self) -> None:
...
+        ]
+
+
+class ExpectedHours(BaseModel):
+    date = models.DateField(help_text=ExpectedHoursConstants.MESSAGE_FOR_CORRECT_DATE_FORMAT.value)
+    value = models.DurationField()
+    user = models.ForeignKey(CustomUser, on_delete=models.CASCADE, related_name="user_expected_hours")
+    project = models.ForeignKey(Project, on_delete=models.CASCADE, related_name="project_expected_hours")
+
+    objects = ExpectedHoursQuerySet.as_manager()
```
[the remark this answers is not in the document]
Comment (Merrique) [reply managers/models.py:86]: changed

File: managers/models.py [hunk 9]
```diff
@@ -59,6 +65,55 @@ def clean(self) -> None:
...
+
+    def clean(self) -> None:
```
[the remark this answers is not in the document]
Comment (Merrique) [reply managers/models.py:88]: Ok. Changed

File: managers/models.py [hunk 10]
```diff
@@ -59,6 +65,55 @@ def clean(self) -> None:
...
+        super().clean()
+        if not isinstance(self.value, timedelta):
+            raise ValidationError(message=ExpectedHoursStrings.EXPECTED_HOURS_WRONG_FORMAT.value)
```
[the remark this answers is not in the document]
Comment (Merrique) [reply managers/models.py:91]: Removed for 'value' and 'date'.

File: managers/models.py [hunk 11]
```diff
@@ -59,6 +65,55 @@ def clean(self) -> None:
...
+        if not isinstance(self.date, date):
+            raise ValidationError(message=ExpectedHoursStrings.EXPECTED_HOURS_WRONG_DATE_FORMAT.value)
+        if self.value < ExpectedHoursConstants.MIN_EXPECTED_HOURS.value:
+            raise ValidationError(message=ExpectedHoursStrings.EXPECTED_HOURS_MIN_VALUE_NOT_EXCEEDED.value)
```
[the remark this answers is not in the document]
Comment (Merrique) [reply managers/models.py:95]: Added validator, Removed check from 'clean'

File: managers/models.py [hunk 12]
```diff
@@ -59,6 +65,55 @@ def clean(self) -> None:
...
+
+        return query_for_date.aggregate(expected_hours_sum=Coalesce(models.Sum("value"), timedelta()))[
+            "expected_hours_sum"
+        ]
+
+
+class ExpectedHours(BaseModel):
+    date = models.DateField(help_text=ExpectedHoursConstants.MESSAGE_FOR_CORRECT_DATE_FORMAT.value)
+    value = models.DurationField()
+    user = models.ForeignKey(CustomUser, on_delete=models.CASCADE, related_name="user_expected_hours")
```
[the remark this answers is not in the document]
Comment (Merrique) [reply managers/models.py:83]: Ok. Changed. I thought it's better to have unique 'related_name'

File: managers/models.py [hunk 13]
```diff
@@ -59,6 +65,55 @@ def clean(self) -> None:
...
+        if self.project.stop_date is not None:
+            if self.date > self.project.stop_date:
+                raise ValidationError(message=ExpectedHoursStrings.EXPECTED_HOURS_DATE_AFTER_PROJECT_FINISHED.value)
+        if self.project.suspended:
+            raise ValidationError(message=ExpectedHoursStrings.EXPECTED_HOURS_FOR_PROJECT_SUSPENDED.value)
+        if (
+            self.user.user_expected_hours.get_expected_hours_sum_for_a_date(self.date, self.pk) + self.value
+            > ExpectedHoursConstants.MAX_EXPECTED_HOURS.value
+        ):
+            raise ValidationError(message=ExpectedHoursStrings.EXPECTED_HOURS_EXCEED_24H_PER_DAY_PER_USER.value)
```
[the remark this answers is not in the document]
Comment (Merrique) [reply managers/models.py:109]: Added 'MaxValueValidator'

File: managers/tests/test_unit_model.py [hunk 14]
```diff
@@ -7,14 +7,145 @@
...
 
+class DataSetUpToTests(BaseModelTestCase):
+    model_class = ExpectedHours
+    required_input = {"date": datetime.now().date(), "user": None, "value": timedelta(hours=8), "project": None}
+
+    def setUp(self):
+        self.user = CustomUser(
+            email="testuser@codepoets.it", password="newuserpasswd", first_name="John", last_name="Doe"
+        )
+        self.user.save()
```
[the remark this answers is not in the document]
Comment (Merrique) [reply managers/tests/test_unit_model.py:31]: changed

File: managers/tests/test_unit_model.py [hunk 15]
```diff
@@ -7,14 +7,145 @@
...
+
+        self.project = Project(
+            name="Test Project", start_date=datetime.now().date(), stop_date=datetime.now().date() + timedelta(days=30)
+        )
+        self.project.save()
```
[the remark this answers is not in the document]
Comment (Merrique) [reply managers/tests/test_unit_model.py:36]: changed

File: managers/tests/test_unit_model.py [hunk 16]
```diff
@@ -7,14 +7,145 @@
...
+        self.project.managers.add(self.user)
+        self.project.save()
```
[the remark this answers is not in the document]
Comment (Merrique) [reply managers/tests/test_unit_model.py:38]: 'save' was unnecessary. Removed

File: managers/tests/test_unit_model.py [hunk 17]
```diff
@@ -7,14 +7,145 @@
...
 from managers.factories import ProjectFactory
+from managers.models import ExpectedHours
 from managers.models import Project
 from users.common.model_helpers import create_user_using_full_clean_and_save
 from users.common.utils import generate_random_string_from_letters_and_digits
+from users.models import CustomUser
 from utils.base_tests import BaseModelTestCase
 
 
+class DataSetUpToTests(BaseModelTestCase):
```
[the remark this answers is not in the document]
Comment (Merrique) [reply managers/tests/test_unit_model.py:23]: Changed

File: managers/tests/test_unit_model.py (hunk 6 above)
Comment (Merrique) [reply managers/tests/test_unit_model.py:43]: changed

File: managers/models.py [hunk 18]
```diff
@@ -59,6 +67,64 @@ def clean(self) -> None:
...
+
+    def check_if_user_works_in_project(self) -> None:
+        if not (
+            self.user.manager_projects.filter(pk=self.project.pk).exists()
+            or self.user.projects.filter(pk=self.project.pk).exists()
+        ):
+            raise ValidationError(message=ExpectedHoursStrings.USER_NOT_IN_THE_PROJECT.value)
```
Comment (kbeker) [inline managers/models.py:116]: All three methods should be private. Second thing is that you need to always make proper order of methods. You need to look at file like you would read for example a book. I'll try to illustrate you example:
```python
def funA():
    funB()
    funC()
    funD()

def funB():
    funE()

def funE():
    print("funE")

def funC():
    print("funC")

def funD():
    print("funD")
```

The same thing you need to do with methods in this file. First need to be `clean()` method and under it all methods which are called inside `clean`

**Review by kbeker — COMMENTED** [review]
Please fix order in `clean` method and it seems that would be all :)

File: managers/models.py (hunk 18 above)
Comment (Merrique) [reply managers/models.py:116]: Ok. I will remember the order of methods in the future :)

**Review by kbeker — APPROVED** [review]

**Review by kbeker — CHANGES_REQUESTED** [review]
Oh, I found that branch name is still invalid. Please create new branch with proper name

**Pull request closed without merging**